waku/v2/protocol/waku_filter_v2/protocol.nim

```diff
@@ -0,0 +1,184 @@
+      cause: "peer has no subscriptions"
+    ))
+
+  ok()
+
+proc subscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
+  if pubsubTopic.isNone() or contentTopics.len() == 0:
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.BAD_REQUEST,
+      cause: "pubsubTopic and contentTopics must be specified"
+    ))
+
+  let filterCriteria = toHashSet(contentTopics.mapIt((pubsubTopic.get(), it)))
+
+  trace "subscribing peer to filter criteria", peerId=peerId, filterCriteria=filterCriteria
+
+  if peerId in wf.subscriptions:
+    var peerSubscription = wf.subscriptions.mgetOrPut(peerId, initHashSet[FilterCriterion]())
+    if peerSubscription.len() + filterCriteria.len() >= MaxCriteriaPerSubscription:
+      return err(FilterSubscribeError(
+        kind: FilterSubscribeErrorKind.SERVICE_UNAVAILABLE,
+        cause: "peer has reached maximum number of filter criteria"
+      ))
+
+    peerSubscription.incl(filterCriteria)
+    wf.subscriptions[peerId] = peerSubscription
+  else:
+    if wf.subscriptions.len() >= MaxSubscriptions:
+      return err(FilterSubscribeError(
+        kind: FilterSubscribeErrorKind.SERVICE_UNAVAILABLE,
+        cause: "node has reached maximum number of subscriptions"
+      ))
+    debug "creating new subscription", peerId=peerId
+    wf.subscriptions[peerId] = filterCriteria
+
+  ok()
+
+proc unsubscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
+  if pubsubTopic.isNone() or contentTopics.len() == 0:
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.BAD_REQUEST,
+      cause: "pubsubTopic and contentTopics must be specified"
+    ))
+
+  let filterCriteria = toHashSet(contentTopics.mapIt((pubsubTopic.get(), it)))
+
+  trace "unsubscribing peer from filter criteria", peerId=peerId, filterCriteria=filterCriteria
+
+  if peerId notin wf.subscriptions:
+    debug "unsubscibing peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.NOT_FOUND,
+      cause: "peer has no subscriptions"
+    ))
+
+  var peerSubscription = wf.subscriptions.mgetOrPut(peerId, initHashSet[FilterCriterion]())
+  # TODO: consider error response if filter criteria does not exist
+  peerSubscription.excl(filterCriteria)
+
+  if peerSubscription.len() == 0:
+    debug "peer has no more subscriptions, removing subscription", peerId=peerId
+    wf.subscriptions.del(peerId)
+  else:
+    wf.subscriptions[peerId] = peerSubscription
+
+  ok()
+
+proc unsubscribeAll(wf: WakuFilter, peerId: PeerID): FilterSubscribeResult =
+  if peerId notin wf.subscriptions:
+    debug "unsubscibing peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError(
+      kind: FilterSubscribeErrorKind.NOT_FOUND,
+      cause: "peer has no subscriptions"
+    ))
```

Comment:
Same here. I would add a convenience wrapper proc for the errors:

```Nim
FilterSubscribeError.notFound(cause="peer has no subscriptions")
```

Reply:
Added in [4b6854c](https://github.com/waku-org/nwaku/pull/1584/commits/4b6854c3d4edc63f7796122a23f91ae8c0d5702f)